In [1]:
!pip install mlxtend

In [2]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori,association_rules

In [3]:
data = pd.read_excel('Onlineretail.xlsx', header=None)

In [4]:
df1 = data.fillna('')

In [5]:
df = df1.drop_duplicates()

In [7]:
transactions = df[0].apply(lambda x: x.split(',')).tolist()

In [79]:

te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df_encoded = pd.DataFrame(te_ary, columns=te.columns_) #df_encoded = pd.DataFrame(te_ary.astype(int), columns=te.columns_)

In [81]:
df_encoded

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False     True               True      False     True        False   
1          False    False              False      False    False        False   
2          False    False              False      False    False        False   
3          False    False              False      False     True        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
5171       False    False              False      False    False        False   
5172       False    False              False      False    False        False   
5173       False    False              False      False    False        False   
5174       False    False              False      False    False        False   
5175       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...   False   
3     False           False      False        False  ...    True   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
5171  False           False      False        False  ...   False   
5172  False           False      False        False  ...    True   
5173  False           False      False        False  ...   False   
5174  False           False      False        False  ...   False   
5175  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
5171           False        False       False             False   
5172           False        False       False             False   
5173           False        False       False             False   
5174           False        False       False             False   
5175           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
5171              False             False  False        False     False  
5172              False              True  False        False     False  
5173              False             False  False        False     False  
5174              False             False  False        False     False  
5175              False             False  False         True     False  

[5176 rows x 120 columns]

In [83]:
df_encoded.to_csv('preprocessed_transactions.csv', index=False)

In [85]:
Data = pd.read_csv('preprocessed_transactions.csv')

In [87]:
Data.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

In [89]:
df = pd.get_dummies(Data, dtype = 'int')
df.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

In [91]:
df.dtypes

 asparagus           bool
almonds              bool
antioxydant juice    bool
asparagus            bool
avocado              bool
                     ... 
whole wheat pasta    bool
whole wheat rice     bool
yams                 bool
yogurt cake          bool
zucchini             bool
Length: 120, dtype: object

In [93]:
# Apriori Algorithm

In [95]:
Frequent_itemsets = apriori(df, min_support = 0.01, use_colnames = True)
Frequent_itemsets

support                               itemsets
0    0.029366                              (almonds)
1    0.011206                    (antioxydant juice)
2    0.045981                              (avocado)
3    0.012558                                (bacon)
4    0.015456                       (barbecue sauce)
..        ...                                    ...
433  0.014683  (mineral water, spaghetti, olive oil)
434  0.016615   (mineral water, spaghetti, pancakes)
435  0.012365     (mineral water, spaghetti, shrimp)
436  0.010819       (soup, mineral water, spaghetti)
437  0.013524   (mineral water, spaghetti, tomatoes)

[438 rows x 2 columns]

In [97]:
rules = association_rules(Frequent_itemsets, metric = 'lift', min_threshold = 1.0)
rules

antecedents                 consequents  \
0              (mineral water)                   (almonds)   
1                    (almonds)             (mineral water)   
2                  (chocolate)                   (avocado)   
3                    (avocado)                 (chocolate)   
4                    (avocado)              (french fries)   
..                         ...                         ...   
835  (mineral water, tomatoes)                 (spaghetti)   
836      (spaghetti, tomatoes)             (mineral water)   
837            (mineral water)       (spaghetti, tomatoes)   
838                (spaghetti)   (mineral water, tomatoes)   
839                 (tomatoes)  (mineral water, spaghetti)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.299845            0.029366  0.011012    0.036727  1.250644   
1              0.029366            0.299845  0.011012    0.375000  1.250644   
2              0.205178            0.045981  0.010240    0.049906  1.085347   
3              0.045981            0.205178  0.010240    0.222689  1.085347   
4              0.045981            0.192620  0.011592    0.252101  1.308800   
..                  ...                 ...       ...         ...       ...   
835            0.034583            0.229521  0.013524    0.391061  1.703817   
836            0.029946            0.299845  0.013524    0.451613  1.506152   
837            0.299845            0.029946  0.013524    0.045103  1.506152   
838            0.229521            0.034583  0.013524    0.058923  1.703817   
839            0.091963            0.085008  0.013524    0.147059  1.729947   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
0                 1.0  0.002207    1.007641       0.286240  0.034608   
1                 1.0  0.002207    1.120247       0.206476  0.034608   
2                 1.0  0.000805    1.004131       0.098935  0.042502   
3                 1.0  0.000805    1.022528       0.082426  0.042502   
4                 1.0  0.002735    1.079531       0.247313  0.051064   
..                ...       ...         ...            ...       ...   
835               1.0  0.005587    1.265282       0.427880  0.053971   
836               1.0  0.004545    1.276752       0.346431  0.042761   
837               1.0  0.004545    1.015873       0.479975  0.042761   
838               1.0  0.005587    1.025864       0.536137  0.053971   
839               1.0  0.005706    1.072750       0.464681  0.082742   

     certainty  kulczynski  
0     0.007583    0.205863  
1     0.107340    0.205863  
2     0.004114    0.136297  
3     0.022032    0.136297  
4     0.073672    0.156141  
..         ...         ...  
835   0.209663    0.224992  
836   0.216763    0.248358  
837   0.015625    0.248358  
838   0.025212    0.224992  
839   0.067816    0.153075  

[840 rows x 14 columns]

In [99]:
rules.sort_values('lift',ascending = False)[0:20]


antecedents                       consequents  \
425                       (olive oil)               (whole wheat pasta)   
424               (whole wheat pasta)                       (olive oil)   
794             (mineral water, milk)                            (soup)   
795                            (soup)             (mineral water, milk)   
299                     (ground beef)                   (herb & pepper)   
298                   (herb & pepper)                     (ground beef)   
735           (mineral water, shrimp)               (frozen vegetables)   
738               (frozen vegetables)           (mineral water, shrimp)   
716    (spaghetti, frozen vegetables)                     (ground beef)   
721                     (ground beef)    (spaghetti, frozen vegetables)   
804                 (spaghetti, milk)                       (olive oil)   
809                       (olive oil)                 (spaghetti, milk)   
719                       (spaghetti)  (ground beef, frozen vegetables)   
718  (ground beef, frozen vegetables)                       (spaghetti)   
797                            (milk)             (soup, mineral water)   
792             (soup, mineral water)                            (milk)   
465                         (burgers)              (french fries, eggs)   
464              (french fries, eggs)                         (burgers)   
773                     (ground beef)        (mineral water, spaghetti)   
768        (mineral water, spaghetti)                     (ground beef)   

     antecedent support  consequent support   support  confidence      lift  \
425            0.087713            0.040572  0.011012    0.125551  3.094525   
424            0.040572            0.087713  0.011012    0.271429  3.094525   
794            0.067813            0.070904  0.012365    0.182336  2.571586   
795            0.070904            0.067813  0.012365    0.174387  2.571586   
299            0.135819            0.066461  0.022798    0.167852  2.525588   
298            0.066461            0.135819  0.022798    0.343023  2.525588   
735            0.033617            0.129830  0.010433    0.310345  2.390394   
738            0.129830            0.033617  0.010433    0.080357  2.390394   
716            0.039026            0.135819  0.012558    0.321782  2.369196   
721            0.135819            0.039026  0.012558    0.092461  2.369196   
804            0.050039            0.087713  0.010240    0.204633  2.332999   
809            0.087713            0.050039  0.010240    0.116740  2.332999   
719            0.229521            0.024536  0.012558    0.054714  2.229911   
718            0.024536            0.229521  0.012558    0.511811  2.229911   
797            0.170015            0.033423  0.012365    0.072727  2.175933   
792            0.033423            0.170015  0.012365    0.369942  2.175933   
465            0.113794            0.046368  0.011399    0.100170  2.160328   
464            0.046368            0.113794  0.011399    0.245833  2.160328   
773            0.135819            0.085008  0.024730    0.182077  2.141885   
768            0.085008            0.135819  0.024730    0.290909  2.141885   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
425               1.0  0.007454    1.097180       0.741925  0.093904   
424               1.0  0.007454    1.252159       0.705471  0.093904   
794               1.0  0.007557    1.136281       0.655593  0.097859   
795               1.0  0.007557    1.129085       0.657774  0.097859   
299               1.0  0.013771    1.121843       0.698989  0.127018   
298               1.0  0.013771    1.315390       0.647056  0.127018   
735               1.0  0.006068    1.261747       0.601893  0.068182   
738               1.0  0.006068    1.050825       0.668443  0.068182   
716               1.0  0.007257    1.274194       0.601386  0.077381   
721               1.0  0.007257    1.058879       0.668744  0.077381  

In [101]:
df = rules[rules.lift>2]
df

antecedents                         consequents  \
298                     (herb & pepper)                       (ground beef)   
299                       (ground beef)                     (herb & pepper)   
376                 (whole wheat pasta)                              (milk)   
377                              (milk)                 (whole wheat pasta)   
418                              (soup)                         (olive oil)   
419                         (olive oil)                              (soup)   
424                 (whole wheat pasta)                         (olive oil)   
425                         (olive oil)                 (whole wheat pasta)   
464                (french fries, eggs)                           (burgers)   
465                           (burgers)                (french fries, eggs)   
548      (chocolate, frozen vegetables)                              (milk)   
553                              (milk)      (chocolate, frozen vegetables)   
622              (chocolate, spaghetti)                         (olive oil)   
627                         (olive oil)              (chocolate, spaghetti)   
655           (frozen vegetables, eggs)                              (milk)   
658                              (milk)           (frozen vegetables, eggs)   
716      (spaghetti, frozen vegetables)                       (ground beef)   
718    (ground beef, frozen vegetables)                         (spaghetti)   
719                         (spaghetti)    (ground beef, frozen vegetables)   
721                       (ground beef)      (spaghetti, frozen vegetables)   
734  (mineral water, frozen vegetables)                            (shrimp)   
735             (mineral water, shrimp)                 (frozen vegetables)   
738                 (frozen vegetables)             (mineral water, shrimp)   
739                            (shrimp)  (mineral water, frozen vegetables)   
757                   (spaghetti, milk)                       (ground beef)   
760                       (ground beef)                   (spaghetti, milk)   
768          (mineral water, spaghetti)                       (ground beef)   
773                       (ground beef)          (mineral water, spaghetti)   
774               (mineral water, milk)                         (olive oil)   
779                         (olive oil)               (mineral water, milk)   
792               (soup, mineral water)                              (milk)   
794               (mineral water, milk)                              (soup)   
795                              (soup)               (mineral water, milk)   
797                              (milk)               (soup, mineral water)   
804                   (spaghetti, milk)                         (olive oil)   
809                         (olive oil)                   (spaghetti, milk)   

     antecedent support  consequent support   support  confidence      lift  \
298            0.066461            0.135819  0.022798    0.343023  2.525588   
299            0.135819            0.066461  0.022798    0.167852  2.525588   
376            0.040572            0.170015  0.013910    0.342857  2.016623   
377            0.170015            0.040572  0.013910    0.081818  2.016623   
418            0.070904            0.087713  0.012944    0.182561  2.081360   
419            0.087713            0.070904  0.012944    0.147577  2.081360   
424            0.040572            0.087713  0.011012    0.271429  3.094525   
425            0.087713            0.040572  0.011012    0.125551  3.094525   
464            0.046368            0.113794  0.011399    0.245833  2.160328   
465            0.113794            0.046368  0.011399    0.100170  2.160328   
548            0.033037            0.170015  0.011592    0.350877  2.063796   
553            0.170015            0.033037  0.011592    0.068182  2.063796   
622            0.055835            0.087713  0.010240    0.183391  2.090819  

In [103]:
df.sort_values('confidence',ascending = False)[0:20]

antecedents                 consequents  \
718    (ground beef, frozen vegetables)                 (spaghetti)   
792               (soup, mineral water)                      (milk)   
548      (chocolate, frozen vegetables)                      (milk)   
655           (frozen vegetables, eggs)                      (milk)   
298                     (herb & pepper)               (ground beef)   
376                 (whole wheat pasta)                      (milk)   
716      (spaghetti, frozen vegetables)               (ground beef)   
735             (mineral water, shrimp)         (frozen vegetables)   
768          (mineral water, spaghetti)               (ground beef)   
757                   (spaghetti, milk)               (ground beef)   
424                 (whole wheat pasta)                 (olive oil)   
464                (french fries, eggs)                   (burgers)   
734  (mineral water, frozen vegetables)                    (shrimp)   
804                   (spaghetti, milk)                 (olive oil)   
622              (chocolate, spaghetti)                 (olive oil)   
418                              (soup)                 (olive oil)   
794               (mineral water, milk)                      (soup)   
774               (mineral water, milk)                 (olive oil)   
773                       (ground beef)  (mineral water, spaghetti)   
795                              (soup)       (mineral water, milk)   

     antecedent support  consequent support   support  confidence      lift  \
718            0.024536            0.229521  0.012558    0.511811  2.229911   
792            0.033423            0.170015  0.012365    0.369942  2.175933   
548            0.033037            0.170015  0.011592    0.350877  2.063796   
655            0.030912            0.170015  0.010626    0.343750  2.021875   
298            0.066461            0.135819  0.022798    0.343023  2.525588   
376            0.040572            0.170015  0.013910    0.342857  2.016623   
716            0.039026            0.135819  0.012558    0.321782  2.369196   
735            0.033617            0.129830  0.010433    0.310345  2.390394   
768            0.085008            0.135819  0.024730    0.290909  2.141885   
757            0.050039            0.135819  0.013910    0.277992  2.046782   
424            0.040572            0.087713  0.011012    0.271429  3.094525   
464            0.046368            0.113794  0.011399    0.245833  2.160328   
734            0.050425            0.098918  0.010433    0.206897  2.091595   
804            0.050039            0.087713  0.010240    0.204633  2.332999   
622            0.055835            0.087713  0.010240    0.183391  2.090819   
418            0.070904            0.087713  0.012944    0.182561  2.081360   
794            0.067813            0.070904  0.012365    0.182336  2.571586   
774            0.067813            0.087713  0.012365    0.182336  2.078793   
773            0.135819            0.085008  0.024730    0.182077  2.141885   
795            0.070904            0.067813  0.012365    0.174387  2.571586   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
718               1.0  0.006926    1.578240       0.565425  0.052000   
792               1.0  0.006682    1.317315       0.559115  0.064712   
548               1.0  0.005975    1.278625       0.533067  0.060545   
655               1.0  0.005370    1.264738       0.521531  0.055838   
298               1.0  0.013771    1.315390       0.647056  0.127018   
376               1.0  0.007013    1.263020       0.525440  0.070727   
716               1.0  0.007257    1.274194       0.601386  0.077381   
735               1.0  0.006068    1.261747       0.601893  0.068182   
768               1.0  0.013184    1.218717       0.582651  0.126108   
757               1.0  0.007114    1.196914       0.538367  0.080899   
424               1.0  0.007454    1.252159       0.705471  0.093904   
464         

## Analysis of the Rules

1. Rule: {olive oil} → {whole wheat pasta}
Support: 0.011 (1.1% of transactions include both olive oil and whole wheat pasta).

Confidence: 0.126 (12.6% of customers who buy olive oil also buy whole wheat pasta).

Lift: 3.095 (Customers are 3.1 times more likely to buy whole wheat pasta if they buy olive oil).

Insight:

Olive oil and whole wheat pasta are strongly associated, likely because they are used together in cooking.

This suggests a health-conscious or cooking-focused customer segment.

2. Rule: {whole wheat pasta} → {olive oil}
Support: 0.011 (1.1% of transactions include both whole wheat pasta and olive oil).

Confidence: 0.271 (27.1% of customers who buy whole wheat pasta also buy olive oil).

Lift: 3.095 (Customers are 3.1 times more likely to buy olive oil if they buy whole wheat pasta).

Insight:

This rule reinforces the strong relationship between olive oil and whole wheat pasta.

Customers who buy whole wheat pasta are highly likely to also buy olive oil, possibly for meal preparation.

3. Rule: {mineral water, milk} → {soup}
Support: 0.012 (1.2% of transactions include mineral water, milk, and soup).

Confidence: 0.182 (18.2% of customers who buy mineral water and milk also buy soup).

Lift: 2.572 (Customers are 2.6 times more likely to buy soup if they buy mineral water and milk).

Insight:

This suggests that customers who purchase mineral water and milk are likely preparing meals that include soup.

This could indicate a preference for convenience or comfort foods.

4. Rule: {soup} → {mineral water, milk}
Support: 0.012 (1.2% of transactions include soup, mineral water, and milk).

Confidence: 0.174 (17.4% of customers who buy soup also buy mineral water and milk).

Lift: 2.572 (Customers are 2.6 times more likely to buy mineral water and milk if they buy soup).

Insight:

This rule highlights a complementary relationship between soup, mineral water, and milk.

Customers may be purchasing these items together for a complete meal.

5. Rule: {ground beef} → {herb & pepper}
Support: 0.023 (2.3% of transactions include both ground beef and herb & pepper).

Confidence: 0.168 (16.8% of customers who buy ground beef also buy herb & pepper).

Lift: 2.526 (Customers are 2.5 times more likely to buy herb & pepper if they buy ground beef).

Insight:

Ground beef and herb & pepper are often used together in cooking .

This suggests a cooking-focused customer segment.

6. Rule: {herb & pepper} → {ground beef}
Support: 0.023 (2.3% of transactions include both herb & pepper and ground beef).

Confidence: 0.343 (34.3% of customers who buy herb & pepper also buy ground beef).

Lift: 2.526 (Customers are 2.5 times more likely to buy ground beef if they buy herb & pepper).

Insight:

This rule reinforces the strong relationship between herb & pepper and ground beef.

Customers who buy herb & pepper are highly likely to also buy ground beef, possibly for meal preparation.

7. Rule: {mineral water, shrimp} → {frozen vegetables}
Support: 0.010 (1.0% of transactions include mineral water, shrimp, and frozen vegetables).

Confidence: 0.310 (31.0% of customers who buy mineral water and shrimp also buy frozen vegetables).

Lift: 2.390 (Customers are 2.4 times more likely to buy frozen vegetables if they buy mineral water and shrimp).

Insight:

This suggests that customers who purchase shrimp and mineral water are likely preparing meals that include frozen vegetables.

This could indicate a preference for convenience or quick meal preparation.

8. Rule: {frozen vegetables} → {mineral water, shrimp}
Support: 0.010 (1.0% of transactions include frozen vegetables, mineral water, and shrimp).

Confidence: 0.080 (8.0% of customers who buy frozen vegetables also buy mineral water and shrimp).

Lift: 2.390 (Customers are 2.4 times more likely to buy mineral water and shrimp if they buy frozen vegetables).

Insight:

This rule highlights a complementary relationship between frozen vegetables, mineral water, and shrimp.

Customers may be purchasing these items together for a complete meal.

9. Rule: {spaghetti, frozen vegetables} → {ground beef}
Support: 0.013 (1.3% of transactions include spaghetti, frozen vegetables, and ground beef).

Confidence: 0.322 (32.2% of customers who buy spaghetti and frozen vegetables also buy ground beef).

Lift: 2.369 (Customers are 2.4 times more likely to buy ground beef if they buy spaghetti and frozen vegetables).

Insight:

This suggests that customers who purchase spaghetti and frozen vegetables are likely preparing meals that include ground beef.

This could indicate a preference for hearty, home-cooked meals.

10. Rule: {ground beef} → {spaghetti, frozen vegetables}
Support: 0.013 (1.3% of transactions include ground beef, spaghetti, and frozen vegetables).

Confidence: 0.092 (9.2% of customers who buy ground beef also buy spaghetti and frozen vegetables).

Lift: 2.369 (Customers are 2.4 times more likely to buy spaghetti and frozen vegetables if they buy ground beef).

Insight:

This rule reinforces the strong relationship between ground beef, spaghetti, and frozen vegetables.

Customers who buy ground beef are likely preparing meals that include spaghetti and frozen vegetables.

## Insights into Customer Purchasing Behavior
    
Cooking-Focused Customers:

Rules like {olive oil} → {whole wheat pasta} and {ground beef} → {herb & pepper} suggest that many customers are purchasing ingredients for cooking meals at home.

These customers are likely health-conscious or prefer home-cooked meals.

Convenience-Oriented Customers:

Rules like {mineral water, milk} → {soup} and {mineral water, shrimp} → {frozen vegetables} indicate that some customers are purchasing items that are easy to prepare or require minimal cooking.

These customers may be looking for quick and convenient meal options.

Complementary Products:

Many rules show strong complementary relationships between products, such as {spaghetti, frozen vegetables} → {ground beef} and {soup} → {mineral water, milk}.

These relationships can be leveraged for cross-selling and product placement strategies.

Seasonal or Event-Based Purchases:

Rules like {burgers} → {french fries, eggs} could indicate purchases for specific occasions.

These patterns can be used to design targeted promotions during specific times.

## Recommendations

Product Placement:

Place complementary products (olive oil and whole wheat pasta, ground beef and herb & pepper) close to each other in the store to encourage additional purchases.

Promotions:

Offer discounts or bundles for frequently co-purchased items (mineral water, milk, and soup).

Targeted Marketing:

Use the rules to create targeted marketing campaigns. send coupons for frozen vegetables to customers who frequently buy shrimp and mineral water.

Inventory Management:

Stock more of frequently co-purchased items together to meet customer demand.

In [ ]:
3. What are Some Limitations or Challenges of Association Rule Mining?
Too Many Rules

The algorithm can generate thousands of rules, making it hard to interpret results.
Solution: Use thresholds like minimum support, confidence, and lift.
Choosing the Right Thresholds

If support is too high, we miss interesting rare associations.
If support is too low, we get too many irrelevant rules.
Handling Large Datasets

Processing large datasets is computationally expensive.
Solution: Use optimized algorithms like Apriori or FP-Growth.
Redundant or Useless Rules

Many rules are trivial (e.g., "Milk → Bread" is expected).
Solution: Use lift or interestingness measures to filter rules.
Correlation vs. Causation

Association rules show correlation, not causation.
Example: "Ice Cream → Sunglasses" does not mean buying ice cream causes sunglasses purchases; they occur together due to summer.